Content.Server/Ninja/Systems/SpaceNinjaSystem.cs

```diff
@@ -0,0 +1,358 @@
+    /// </summary>
+    public int Download(EntityUid uid, List<string> ids)
+    {
+        if (!GetNinjaRole(uid, out var role))
+            return 0;
+
+        var oldCount = role.DownloadedNodes.Count;
+        role.DownloadedNodes.UnionWith(ids);
+        var newCount = role.DownloadedNodes.Count;
+        return newCount - oldCount;
+    }
+
+    /// <summary>
+    /// Gets a ninja's role using the player's mind
+    /// </summary>
+    public bool GetNinjaRole(Mind.Mind? mind, [NotNullWhen(true)] out NinjaRole? role)
+    {
+        role = null;
+        if (mind == null)
+            return false;
+
+        foreach (var r in mind.AllRoles)
+        {
+            if (r is NinjaRole ninja)
+            {
+                role = ninja;
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /// <summary>
+    /// Gets a ninja's role using the player's entity id
+    /// </summary>
+    public bool GetNinjaRole(EntityUid uid, [NotNullWhen(true)] out NinjaRole? role)
+    {
+        role = null;
+        if (!TryComp<MindContainerComponent>(uid, out var mind))
+            return false;
+
+        return GetNinjaRole(mind.Mind, out role);
+    }
+
+    /// <summary>
+    /// Returns the global ninja gamerule config
+    /// </summary>
+    public NinjaRuleComponent RuleConfig()
+    {
+        var rule = EntityQuery<NinjaRuleComponent>().FirstOrDefault();
+        if (rule != null)
+            return rule;
+
+        _gameTicker.StartGameRule("Ninja", out var ruleEntity);
+        return Comp<NinjaRuleComponent>(ruleEntity);
+    }
+
+    /// <summary>
+    /// Update the alert for the ninja's suit power indicator.
+    /// </summary>
+    public void SetSuitPowerAlert(EntityUid uid, SpaceNinjaComponent? comp = null)
+    {
+        if (!Resolve(uid, ref comp, false) || comp.Deleted || comp.Suit == null)
+        {
+            _alerts.ClearAlert(uid, AlertType.SuitPower);
+            return;
+        }
+
+        if (GetNinjaBattery(uid, out var _, out var battery))
+        {
+             var severity = ContentHelpers.RoundToLevels(MathF.Max(0f, battery.CurrentCharge), battery.MaxCharge, 8);
+            _alerts.ShowAlert(uid, AlertType.SuitPower, (short) severity);
+        }
+        else
+        {
+            _alerts.ClearAlert(uid, AlertType.SuitPower);
+        }
+    }
+
+    /// <summary>
+    /// Get the battery component in a ninja's suit, if it's worn.
+    /// </summary>
+    public bool GetNinjaBattery(EntityUid user, [NotNullWhen(true)] out EntityUid? uid, [NotNullWhen(true)] out BatteryComponent? battery)
+    {
+        if (TryComp<SpaceNinjaComponent>(user, out var ninja)
+            && ninja.Suit != null
+            && _powerCell.TryGetBatteryFromSlot(ninja.Suit.Value, out uid, out battery))
+        {
+            return true;
+        }
+
+        uid = null;
+        battery = null;
+        return false;
+    }
```

Comment:
This should probably just be a generic "PoweredClothingComponent" or something instead of being ninja specific and split out.